src/storage/accessors/DataAccessor.ts

```diff
@@ -0,0 +1,81 @@
+import type { Readable } from 'stream';
+import type { Patch } from '../../ldp/http/Patch';
+import type { Representation } from '../../ldp/representation/Representation';
+import type { RepresentationMetadata } from '../../ldp/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+
+/**
+ * A DataAccessor is the building block closest to the actual data storage.
+ * It should not worry about most Solid logic, most of that will be handled before it is called.
+ * There are a few things it still needs to do, and it is very important every implementation does this:
+ *  * If the input identifier ends with a slash, it should be assumed the identifier is targeting a container.
+ *  * Similarly, if there is no trailing slash it should assume a data resource.
+ *  * It should always throw a NotFoundHttpError if it does not have data matching the input identifier.
+ *  * The only exception is the `getNormalizedMetadata` function where both versions of the identifier
+ *    need to be checked (both with and without slash).
+ *  * DataAccessors are responsible for generating the relevant containment triples for containers.
+ */
+export interface DataAccessor {
+  /**
+   * Should throw an UnsupportedHttpError if the DataAccessor does not support storing the given Representation.
+   * @param representation - Incoming Representation.
+   *
+   * @throws UnsupportedHttpError
+   * If it does not support the incoming data.
+   */
+  canHandle: (representation: Representation) => Promise<void>;
+
+  /**
+   * Returns a data stream stored for the given identifier.
+   * It can be assumed that the incoming identifier will always correspond to a data resource.
+   * @param identifier - Identifier for which the data is requested.
+   */
+  getData: (identifier: ResourceIdentifier) => Promise<Readable>;
+
+  /**
+   * Returns the metadata corresponding to the identifier.
+   * @param identifier - Identifier for which the metadata is requested.
+   */
+  getMetadata: (identifier: ResourceIdentifier) => Promise<RepresentationMetadata>;
+
+  /**
+   * Returns the metadata matching the identifier, ignoring the presence of a trailing slash or not.
+   * This is used to support the following part of the spec:
+   * "If two URIs differ only in the trailing slash, and the server has associated a resource with one of them,
+   *  then the other URI MUST NOT correspond to another resource."
+   * @param identifier - Identifier that needs to be checked.
+   */
+  getNormalizedMetadata: (identifier: ResourceIdentifier) => Promise<RepresentationMetadata>;
```

Comment:
Mmm, why is that in the accessor contract? Seems like a more high-level thing to me.

Reply:
The reason for this one being in the DataAccessor is performance reasons. Specifically a SPARQL backend can do this with 1 query, but would need 2 queries if it is encoded in the DataAccessorBasedStore. So it's a trade-off. Can always move it back if you don't think the trade-off is worth it.

Reply:
Yeah, the accessor contract is there for simplicity. So would not add this. Performance can be an additional interface if needed.